### DISCUSSION TITLE
Switch RetroQuest's login mechansims to use user-based logins

### DISCUSSION THREAD

**Discussion by tsatam** [FordLabs/retroquest#454] [RFCs]
# Summary

Change RetroQuest to use user-based logins tied to an identity provider, which grants them access to a team.

# Motivation

RetroQuest was originally designed with full anonymity as its highest priority: there would be no individual user accounts so no actions could be traced to any individual, and all users would have the same access to everything on their team boards.

As a result, RetroQuest uses a single teamname/password credential for each team board, generated and managed by RetroQuest, and all users on that team board will use that credential to log in.

# Current Concerns

As our usage of RetroQuest has grown larger and more diverse, this model has caused many issues:

## Lost credentials

Users who fully forget their credentials can no longer access their board. Because we have no administrative access to the data for teams, we have no way to reset their password for them, or offer a mechanism for them to reset the password themselves.

## Access management

Some teams will reuse the same team board across multiple retrospective sessions. Users who should no longer have access to a team board cannot be removed from the board. We initially designed RetroQuest for single-use boards, or for the board to only be reused used within a short iteration period, but users have preferred to keep the same board running long-term for their team.

## Reducing our vulnerable surface area for attacks

We have to manage credentials ourselves. This could potentially be a point of failure/weak link in security.

## Poor user experience

Users have to remember "yet another password" in order to use RetroQuest alongside their other tools. Users who use multiple separate retro boards have to remember multiple passwords. 

# Design

The proposed solution here is to implement user logins to RetroQuest using single-sign-on:

- Users would sign in with a selected identity provider, and can create or be added to teams.
- Each team would have a set of retros (formerly boards), and a single set of action items shared across all retros.
- All actions performed within a retro will still be anonymous. No actions can be traced back to individual users.
- All users within a team will have the same access/permissions to actions within the team.

The plan is to allow RetroQuest to use an OAuth OIDC-backed provider for user logins. In corporate environments, for example, this can be hooked up to whatever system is used for employee sign-ins, while local development can be hooked up to a multitude of providers running either in the cloud or locally.

Our intent is to figure out an easy solution for interested consumers of RetroQuest to run the app with a locally hosted OAuth provider for testing and light production usage.

## Preserving anonymity

RetroQuest is designed to be a safe environment where people can discuss topics freely as a part of their team, without worrying about anything being traced back to them in particular. Additionally, RetroQuest is also designed with anonymity in mind, so there are no elevated permissions for anyone, whether that be "facilitators" of a retro, or central administrators. 

As RetroQuest has developed, there has been more of a need for central administration, primarily around access controls.

We want to ensure this change still preserves our original goals for RetroQuest while allowing for this change. As a result, we do not intend anything within an actual retro to be tied to user accounts, they will remain anonymous just as they always have. We do not intend to add any user tracking for what thoughts are entered, what action items are created, or for who has voted on topics. 

We will also not be adding in "facilitator" level administrative features on retros, allowing any user with valid access to their team to perform any action on the retro. 

There have been explicit requests to add user-level features such as the above, but we will not support those requests directly with this change, and will pursue alternative mechanisms for achieving these asks without violating the core values of RetroQuest as a separate exercise. 

This does open the door for administrative-level management of what teams exist and who is a member of those teams. We currently do not consider that to be information that must be kept private within a team, but we're open to further discussion on whether or not this needs to be locked down as well. Depending on what level of control we provide to users (should users be able to delete their team?), this level of administration may not even be necessary. In any case, this specific functionality would be out of scope for this effort. 

## What about existing content?

This will be a breaking change. Our intent is to "start the slate clean" since the access controls are different than the old model. We plan to offer an "Import from Legacy RetroQuest" option where a user can enter the teamname and password for an old RetroQuest board, and import the contents from that board into a new retro on a team.

## Data Model

This data model is WIP, and only contains the minimum amount of information necessary for this specific change. Any
future enhancements should be done as separate efforts.

### Team

A team represents a shared space for all users belonging to the same unit. It will be the defining factor for access
controls for everything within it: all users within a team will have access to everything within it, and all users not
in that team will have access to nothing within it.

| Field | Type |
|-------|------|
| id    | id   |
| name  | name |

### User

A user represents a single person using RetroQuest, whose identity is backed by a third-party identity provider. A user
can belong to multiple teams.

| Field | Type                    |
|-------|-------------------------|
| id    | id (from OIDC provider) |
| teams | [Team]                  |

### Retro (formerly Board)

A retro represents a single instance of a retrospective. A team will have multiple retros within it.

| Field     | Type    |
|-----------|---------|
| id        | id      |
| team      | Team    |
| name      | name    |
| completed | boolean |

Additional fields or related domain objects listed below will mostly match the existing model for boards in RetroQuest.

### Topic

A topic represents a single grouping of thoughts in a retro. These are represented as columns in the UI.

Each retro will have a set of three topics, allowing different retros within the same team to have different definitions
for their topics.

NOTE: This representation of topics, as well as the implementation of columns in existing RetroQuest, has the potential
to support a different number of topics than three. However, implementing the ability to change this number is out of
scope for this change.

| Field    | Type                        |
|----------|-----------------------------|
| id       | id                          |
| retro    | Retro                       |
| name     | name                        |
| color    | Color                       |
| position | zero-indexed position (int) |

### Thought

A thought represents a single idea/discussion point during a retrospective, belonging to a specific Topic. This
definition is mostly unchanged from the existing model.

| Field     | Type    |
|-----------|---------|
| id        | id      |
| retro     | Retro   |
| topic     | Topic   |
| message   | text    |
| votes     | int     |
| completed | boolean |

### Action Item

Represents a single takeaway activity that was agreed-upon during a retrospective. Unlike the old RetroQuest model,
Action Items will belong to the entire team rather than to a specific retro, and all retros will interact with the same
set of action items.

NOTE: Despite users now being a top level concept in RetroQuest, we still will not be constraining assignees to only be
RetroQuest users. We may be able to provide enhancements such as autocomplete, but that is out of scope for this change.

| Field     | Type    |
|-----------|---------|
| id        | id      |
| team      | Team    |
| task      | text    |
| completed | boolean |
| assignee  | text    |

# Benefits

- User management would largely be managed externally: a corporate-managed OIDC provider would automatically handle the creation and deletion of relevant user accounts on RetroQuest's behalf for organizations using RetroQuest that require secure access for employees only, while public-facing deployments can utilize any number of public OIDC handlers to facilitate user logins without the toil of managing user accounts themselves. 
- Users would not need to remember another password to access RetroQuest; it would use the same login mechanisms as their other tools
- Teams would provide an organizational construct not present in the existing usage of RetroQuest

# Drawbacks

- This is a breaking change, requiring users to manually migrate their content into a new working instance. We hope to make this as painless as possible for teams to achieve.
- This is an expensive change, and will require a fair amount of work to achieve. We hope the plan outlined below reduces this cost as much as possible, and are open to further discussion around reducing this further.

# Alternatives

## Do nothing, maintain the existing product as-is

The existing drawbacks would continue to exist

## Implement some sort of mechanism to reset passwords for boards

Implement some sort of flow (e.g. a reset password email, or a one-time token provided on board creation) to allow users to reset the credentials for their board.

This would, on the surface, be a simpler option to reduce the access management burden for users. However, we do not see an easy path forward to make this safe and not itself prone to malicious actors (who gets to reset the password on a board?).

If, for example, we asked users who create a board to provide an email address that the reset password flow would operate on, or provided the user a reset token, this change would do nothing for the existing users of RetroQuest who have already created a board without this value.

Additionally, it does not solve any of the user experience concerns around access management, and would greatly increase our implementation complexity.

# Plan

We hope to keep this change as small and focused as possible, given how large the impact is going to be. As a result, our intent is to make this primarily a backend change; there will not be any significant differences to the user interface for the actual retro board itself.

We will still need to design a user interface for team management. Since the concept of archiving retros will largely be replaced by teams containing multiple retros, we can reuse that existing UI for switching between retros on a team.

The implementation of this will end up looking incredibly similar to our sister product [PeopleMover](https://github.com/FordLabs/peoplemover), and we hope to hit the ground running by reusing large parts of its existing implementation.

**Comment by LowBudgetMan** [comment]
I think this makes a lot of sense to implement. Adding support for OAuth venders instead of our rolled user controls will make it a lot easier for team management down the line. In regards to the migration though, would this mean anyone running the previous versions of RetroQuest would need to keep their old instances up indefinitely for the import to work? Would it be better to be able to upload a file that was exported from an old version of RetroQuest, so it's not dependent on a running application?

**Reply by tsatam** [reply]
I'm not sure how we exactly want to handle that just yet. Options include:

1. Deployers will have to run both a legacy version and current version of RetroQuest in order to facilitate migrating data over. Current RetroQuest will interact with legacy RetroQuest's API in order to retrieve data. 
2. Deployers keep database tables for the legacy version and the current version co-hosted in the same database, so the "import from legacy RetroQuest" option just reaches in to existing tables to grab the data. 
3. Same as option 2, but deployers manually export the data from legacy RetroQuest into new tables in current RetroQuest, so we can keep the old table structure out of our data definitions. 
4. Same as option 3, but we make RetroQuest depend on a non-database export of data stored in some location (your suggestion)
5. Deployers ask users to upload a CSV export from legacy RetroQuest into new RetroQuest. This is probably the cleanest maintenance option long-term, but IMO provides the worst user experience. This would also, in all practicality, require deployers to run both versions of RetroQuest for a period of time to facilitate users performing this migration. 

For our deployment internally at Ford, option 1 seems like the easiest option, but I'm completely open to ideas to make this more amenable to other instances of RetroQuest as well.

**Reply by LowBudgetMan** [reply]
I like option 5 the best for long term.

**Reply by tsatam** [reply]
The current CSV export in RetroQuest only exports the current state of the active retro board. I think if we are to go with option 5, we ought to either make the CSV export more customizable, or have it export all archived content as well.

**Comment by aclift19** [comment]
I like this plan Waaaay better than trying to roll our own password reset and managing all of that. Seems a lot more sustainable in the long run.

I also love the idea of having the UI for people to see what teams they are a part of. I for one would love that.

**Comment by sapphon** [comment]
Here are two alternatives to ADFSQuest.  They have attributes in common:

1.  Each would cost a fraction of what ADFSQuest would to implement,
2. Each would leave one problem stated in the above RFC at least partially unsolved,
3. Each would preserve RetroQuest's tradition of user anonymity.

* Provide a secret on board creation that allows password reset.  The disadvantage here is that one still has to create a new board when the board owner themselves becomes *non grata*.  For an example of this in action, see [whenisgood.net](whenisgood.net)'s admin system for polls.

* Allow anyone with a board's password to "fork" it, creating a copy with a new password; teams can fork when someone with the current password becomes *non grata*.  This comes with the advantage that participants in past retros aren't locked out of retro content they contributed to - only the content of future retros conducted using new forks.  It comes with the disadvantage of having to disseminate the new password for / grant access to each fork.  For an example of this in action, see the website we're using now.
